homeassistant/components/airthings_ble/sensor.py

```diff
@@ -107,9 +115,70 @@
 }
 
 
+@callback
+def async_migrate(hass: HomeAssistant, address: str, sensor_name: str):
+    """Migrate entities to new unique ids (with BLE Address)."""
+
+    ent_reg = entity_async_get(hass)
+
+    device_registry = device_async_get(hass)
+    entity_registry = entity_async_get(hass)
+
+    device = device_registry.async_get_device(
+        connections={(CONNECTION_BLUETOOTH, address)}
+    )
+
+    if not device:
+        return
+
+    entities = async_entries_for_device(
+        entity_registry,
+        device_id=device.id,
+        include_disabled_entities=True,
+    )
+
+    filtered_entities = (
+        entity for entity in entities if entity.unique_id.endswith(sensor_name)
+    )
+
+    unique_ids: dict[str, str] = {}
+
+    for entity in filtered_entities:
+        if entity.unique_id.startswith(address):
+            unique_ids["v3"] = entity.entity_id
+        elif "(" in entity.unique_id:
+            unique_ids["v2"] = entity.entity_id
+        else:
+            unique_ids["v1"] = entity.entity_id
+
+    def _migrate_unique_id(entity_id: str, new_unique_id: str):
+        ent_reg.async_update_entity(entity_id=entity_id, new_unique_id=new_unique_id)
+        _LOGGER.debug(
+            "Migrated entity '%s' to unique id '%s'", entity_id, new_unique_id
+        )
+
+    if unique_ids.get("v3"):
+        # Already has the newest unique id format
+        return
+
+    new_unique_id = f"{address}_{sensor_name}"
+
+    # Try to migrate the pre 2023.9.0 unique id first
+    if entity_id := unique_ids.get("v1"):
+        _migrate_unique_id(
+            entity_id=entity_id,
+            new_unique_id=new_unique_id,
+        )
+
+    # If pre 2023.9.0 unique id is not found, try to migrate the unique id
+    # introduced in 2023.9.0
+    elif entity_id := unique_ids.get("v2"):
+        _migrate_unique_id(entity_id=entity_id, new_unique_id=new_unique_id)
```

Comment:
```suggestion
@callback
def async_migrate(hass: HomeAssistant, address: str, sensor_name: str) -> None:
    """Migrate entities to new unique ids (with BLE Address)."""
    ent_reg = entity_async_get(hass)
    unique_id_trailer = f"_{sensor_name}"
    new_unique_id = f"{address}{unique_id_trailer}"
    if ent_reg.async_get_entity_id(DOMAIN, Platform.SENSOR, new_unique_id):
        # New unique id already exists
        return

    dev_reg = device_async_get(hass)
    if not (
        device := dev_reg.async_get_device(
            connections={(CONNECTION_BLUETOOTH, address)}
        )
    ):
        return

    entities = async_entries_for_device(
        ent_reg,
        device_id=device.id,
        include_disabled_entities=True,
    )
    matching_reg_entry: RegistryEntry | None = None
    for entry in entities:
        if entry.unique_id.endswith(unique_id_trailer):
            matching_reg_entry = entry
            break

    if not matching_reg_entry or matching_reg_entry.unique_id.startswith(address):
        # Already has the newest unique id format
        return

    entity_id = matching_reg_entry.entity_id
    ent_reg.async_update_entity(entity_id=entity_id, new_unique_id=new_unique_id)
    _LOGGER.debug("Migrated entity '%s' to unique id '%s'", entity_id, new_unique_id)
```

I think this can be simplified a bit

Reply:
Nice! But are we guaranteed that the oldest entry will be found first and used? If I change the order in the test to add the newest format first, then the old format ([these lines](https://github.com/home-assistant/core/pull/99832/files?diff=unified&w=0#diff-e05d3e97d19748dccce281e7885ea7393e73d7471e45e09d829a8b19ae8b2a3fR120-R134)), the test will fail. The difference between my code and yours is that my code will try to find the oldest (v1), then the 2023.9.0 format (v2). Your code doesn't seem to take care of that. If we are guaranteed that the oldest (v1) format will be found first and used, this will work, and we can skip some logic

Reply:
It won't currently but it can be adjusted to not `break` in and let it keep going so the oldest one wins there.
```python
    for entry in entities:
        if entry.unique_id.endswith(unique_id_trailer):
            matching_reg_entry = entry
            break
```

I guess we need a test for this case

Reply:
Updated the code and changed the logic a bit:

If `entry.unique_id` does _not_ contain a `(` it means it's the format from before 2023.9.0, and we will try to migrate that one.

If `matching_reg_entry` is None we will still save whatever we find of `unique_id`. The result will be the v1 format if it exists.

Also removed the check to see if it finds a entry with the newest format, because that was already done earlier:
`if ent_reg.async_get_entity_id(DOMAIN, Platform.SENSOR, new_unique_id):`